Day4/src/pipeline/extract_employe_data.py

```diff
@@ -0,0 +1,355 @@
+    """
+    update2="""
+    UPDATE temporary_attendance_table ta
+    SET on_call_hour = b.on_call_hour
+    FROM temporary_on_call b
+    WHERE b.employee_id=ta.employee_id AND b.punch_apply_date =ta.shifT_date;
+    """
+    cur.execute(update)
+    cur.execute(update2)
+    con.commit()
+
+def add_department_id(cur,con):
+    alter="ALTER TABLE temporary_attendance_table ADD department_id VARCHAR(250);"  
+    cur.execute(alter)
+    con.commit()
+    update="""
+    UPDATE temporary_attendance_table ta
+    SET department_id = e.department_id
+    FROM raw_employee e
+    WHERE e.employee_id = ta.employee_id;
+    """
+    cur.execute(update)
+    con.commit()
+
+def create_absent_count(cur,con):
+    delete="DROP TABLE IF EXISTS temporary_absent_count;"
+    cur.execute(delete)
+    con.commit()
+    create="""
+    CREATE TABLE temporary_absent_count AS 
+    SELECT shift_date,department_id,COUNT(attendance)
+    FROM temporary_attendance_table
+    GROUP by shift_date,department_id
+    ORDER by 1 ASC, 2 ASC;
+    """
+    cur.execute(create)
+    con.commit()
+   
+def update_table_with_absent_count(cur,con):
+    alter="ALTER TABLE temporary_attendance_table ADD num_teammates_absent NUMERIC DEFAULT 0;"
+    cur.execute(alter)
+    con.commit()
+    update="""
+    UPDATE temporary_attendance_table ta
+    SET num_teammates_absent= ac.count
+    FROM temporary_absent_count ac
+    WHERE ac.department_id=ta.department_id AND ac.shift_date=ta.shift_date;
+    """
+    cur.execute(update)
+    con.commit()
+
+def create_table_timesheet(cur,con):
+    delete="DROP TABLE IF EXISTS transformation_timesheet;"
+    cur.execute(delete)
+    con.commit()
+
+    create_table="""
+    CREATE TABLE transformation_timesheet AS
+    SELECT employee_id,department_id,shift_start_time,shift_end_time,shift_date,shift_type,hours_worked,
+    CASE WHEN hours_worked<3 THEN 'ABSENT' ELSE attendance END as attendance,
+    has_taken_break,break_hour,was_charge,charge_hour,was_on_call,on_call_hour,num_teammates_absent
+    FROM temporary_attendance_table;
+    """
+    cur.execute(create_table)
+    con.commit()
+
+def main():
+    con =connect()
+    cur=con.cursor()
+    extract_employee_data(cur,con)
+    extract_timesheet_data(cur,con)
+    clean_hours_worked(cur,con)
+    create_temporary_punch_details(cur,con)
+    create_temporary_shift_time(cur,con)
+    create_temporary_tables_for_hours_worked(cur,con)
+    create_table_add_hours(cur,con)
+    insert_data_to_one_table(cur,con)
+    create_attendance_record(cur,con)
+    create_full_attendance(cur, con)
+    create_temporary_break(cur,con)
+    update_table_with_break(cur,con)
+    create_temporary_charge(cur,con)
+    update_table_with_charge(cur,con)
+    create_temporary_on_call(cur,con)
+    update_table_with_on_call(cur,con)
+    add_department_id(cur,con)
+    create_absent_count(cur,con)
+    update_table_with_absent_count(cur,con)
```

Comment:
All the calculations in these procedures seem repeated. I think this can be done with a single procedure.

Reply:
Yes, I will do it in BEGIN END block